resources/type-system/flow-analysis.md

```diff
@@ -270,95 +262,105 @@ We also make use of the following auxiliary functions:
-     from the try block in any subsequent code (since if any subsequent code is
-     executed, the try block must have completed normally).  We only choose to
-     do so if the last entry is more precise.  (TODO: is this the right thing to
-     do here?).
-   - `s3 = s2`
-     - The set of types of interest is the set of types of interest in the
-       finally block.
-   - `a3 = a2 || a1`
-     - A variable is definitely assigned after the finally block if it is
-       definitely assigned by the try block or by the finally block (note that
-       code after the finally block will only be executed if the try block
-       completes normally).
-   - `u3 = u2`
-     - A variable is definitely unassigned if it is definitely unassigned in the
-       model of the finally block (note that the finally block is analyzed using
-       the join of the model from before the try block and after the try block,
-       and so the absence of any assignments that may have occurred in the try
-       block is already modelled here).
-
-- `restrict(MB, MF, N)`, where `MB` and `MF` are flow models and `N` is a set of
-  variables assigned in the finally clause, models the flow of information
-  through a try/finally statement, and is defined as follows:
-
-  - We define `restrict(MB, MF, N)` to be `M3 = FlowModel(r3, VI3)` where:
-    - `MB = FlowModel(rb, VIB)`
-    - `MF = FlowModel(rf, VIF))`
-    - `pop(rb) = pop(rf) = r0` for some `r0`
-    - `r3` is `push(r0, top(rb) && top(rf))`
-    - `b` is true if `v` is in `N` and otherwise false
-    - `VI3` is the map which maps each variable `v` in the domain of `VIB` and
-      `VIF` to `restrictV(VIB(v), VIF(v), b)`.
+  For brevity, we will sometimes extend `join` to more than two arguments in the
+  obvious way.  For example, `join(M1, M2, M3)` represents `join(join(M1, M2),
+  M3)`, and `join(S)`, where S is a set of models, denotes the result of folding
+  all models in S together using `join`.
+
+  Simiarly, if L is a non-empty list of flow models, `join(...L)` represents the
+  result of folding the list using `join`.
+
+- `rebasePromotedTypes(p1, p2)`, where `p1` and `p2` are promotion chains,
+  computes a promotion chain containing promotions from both `p1` and `p2`,
+  discarding promotions if necessary to avoid violating the promotion chain
+  requirement. It is defined as `p3`, where:
+  - If `p2` is empty, then `p3 = p1`.
+  - Otherwise:
+    - Let `T2` be the last type in `p2`.
+    - Let `p1'` be a promotion chain obtained by deleting any elements `T` from
+      `p1` that do not satisfy `T <: T2`.
+    - Let `p3 = [...p2, ...p1']`.
+
+- `attachFinallyV(VM1, VM2, VM3)`, where `VM1`, `VM2`, and `VM3` are variable
+  models, represents the state of a variable model after a `try/finally`
+  statement, where `VM1` is the state after the `try` block, `VM2` is the state
+  before the `finally` block, and `VM3` is the state after the `finally`
```

Comment:
And the state before the `finally` block is the same as the state before the entire `try/finally` statement (because the `try` block can throw at any point)?

Also, the state after the `finally` block, does that mean at the end of the finally block? Presumably the whole point of this operation is that it computes the stat after the `try/finally` statement as a whole, which is also right after the `finally` block?

Perhaps add a bit of commentary text here, to help readers.

Reply:
Upon further reflection, I think a lot of the confusion here was because the `attachFinallyV` function was described first, whereas the explanation of the meaning of the various arguments was in the definition of `attachFinally`, which followed. I've reversed the order of the two to make things clearer. Also, as we discussed in our meeting, I've renamed the arguments to `attachFinally` and `attachFinallyV` so that their meaning is clear from their names.